scipio/src/channels/local_channel.rs

```diff
@@ -0,0 +1,586 @@
+    ///     let h = Local::local(async move {
+    ///         assert_eq!(receiver.next().await.unwrap(), 0);
+    ///     }).detach();
+    ///     sender.send(0);
+    ///     drop(sender);
+    ///     h.await;
+    /// });
+    /// ```
+    ///
+    /// [`LocalChannel`]: struct.LocalChannel.html
+    pub fn new_unbounded() -> (LocalSender<T>, LocalReceiver<T>) {
+        Self::new(ChannelCapacity::Unbounded)
+    }
+
+    /// Creates a new [`LocalChannel`] with capacity limited to the `size` argument
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::{LocalExecutor, Local};
+    /// use scipio::channels::LocalChannel;
+    /// use futures::stream::StreamExt;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async move {
+    ///     let (sender, mut receiver) = LocalChannel::new_bounded(1);
+    ///     assert_eq!(sender.has_room(), true);
+    ///     sender.send(0);
+    ///     assert_eq!(sender.has_room(), false);
+    ///     receiver.next().await.unwrap();
+    ///     assert_eq!(sender.has_room(), true);
+    /// });
+    /// ```
+    ///
+    /// [`LocalChannel`]: struct.LocalChannel.html
+    pub fn new_bounded(size: usize) -> (LocalSender<T>, LocalReceiver<T>) {
+        Self::new(ChannelCapacity::Bounded(size))
+    }
+}
+
+struct SendWaiter<T> {
+    channel: LocalChannel<T>,
+}
+
+impl<T> Future for SendWaiter<T> {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        if self.channel.has_room() {
+            Poll::Ready(())
+        } else {
+            let mut state = self.channel.state.borrow_mut();
+            state.send_waiters.push(cx.waker().clone());
+            Poll::Pending
+        }
+    }
+}
+
+impl<T> LocalSender<T> {
+    /// Sends data into this channel.
+    ///
+    /// It returns [`io::Error`] [`BrokenPipe`] if the receiver is destroyed.
+    /// It returns [`io::Error`] [`Other`] if this is a bounded channel that has no more capacity
+    ///
+    /// # Examples
+    /// ```
+    /// use scipio::{LocalExecutor, Local};
+    /// use scipio::channels::LocalChannel;
+    /// use futures::stream::StreamExt;
+    ///
+    /// let ex = LocalExecutor::make_default();
+    /// ex.run(async move {
+    ///     let (sender, mut receiver) = LocalChannel::new_bounded(1);
+    ///     sender.send(0);
+    ///     sender.send(0).unwrap_err(); // no more capacity
+    ///     receiver.next().await.unwrap(); // now we have capacity again
+    ///     drop(receiver); // but because the receiver is destroyed send will err
+    ///     sender.send(0).unwrap_err();
+    /// });
+    /// ```
+    ///
+    /// [`BrokenPipe`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.BrokenPipe
+    /// [`Other`]: https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.Other
+    /// [`io::Error`]: https://doc.rust-lang.org/std/io/struct.Error.html
+    pub fn send(&self, item: T) -> io::Result<()> {
```

Comment:
Using io::Result here is problematic, because it won’t be possible to recover the item on error. Idiomatic Rust solution here is to return a type similar to Result<(), T>.

Reply:
right, that makes sense.